x/params/keeper_upgrade.go

```diff
@@ -0,0 +1,144 @@
+package params
+
+import (
+	"fmt"
+
+	"github.com/okex/exchain/libs/cosmos-sdk/store/prefix"
+	sdk "github.com/okex/exchain/libs/cosmos-sdk/types"
+	"github.com/okex/exchain/x/params/types"
+)
+
+var (
+	upgradeInfoPreifx = []byte("upgrade")
+)
+
+// ClaimReadyForUpgrade tells Keeper that someone has get ready for the upgrade.
+// cb could be nil if there's no code to be execute when the upgrade is take effective.
+// NOTE: for the same name, if this method is called more than once time, the callback of
+// latest call will cover all of previous ones.
+func (keeper *Keeper) ClaimReadyForUpgrade(name string, cb func(types.UpgradeInfo)) {
+	if _, ok := keeper.upgradeReadyMap[name]; ok {
+		keeper.logger.Error("more than one guys ready for the same upgrade, the front one will be cover", "upgrade name", name)
+	}
+	keeper.upgradeReadyMap[name] = cb
+}
+
+func (keeper *Keeper) IsUpgradeEffective(ctx sdk.Context, name string) bool {
+	_, err := keeper.GetEffectiveUpgradeInfo(ctx, name)
+	return err == nil
+}
+
+func (keeper *Keeper) GetEffectiveUpgradeInfo(ctx sdk.Context, name string) (types.UpgradeInfo, error) {
+	info, err := keeper.readUpgradeInfo(ctx, name)
+	if err != nil {
+		return types.UpgradeInfo{}, err
+	}
+
+	if !isUpgradeEffective(ctx, info) {
+		keeper.Logger(ctx).Debug("upgrade is not effective", "name", name)
+		return types.UpgradeInfo{}, fmt.Errorf("upgrade '%s' is not effective", name)
+	}
+
+	keeper.Logger(ctx).Debug("upgrade is effective", "name", name)
+	return info, nil
+}
+
+func (keeper *Keeper) queryReadyForUpgrade(name string) (func(types.UpgradeInfo), bool) {
+	cb, ok := keeper.upgradeReadyMap[name]
+	return cb, ok
+}
+
+func (keeper *Keeper) readUpgradeInfo(ctx sdk.Context, name string) (types.UpgradeInfo, error) {
+	info, exist := keeper.readUpgradeInfoFromCache(name)
```

Comment:
Not safe in parallel-tx , Please refer to this way of writing  https://github.com/okx/exchain/blob/bd642b287d58bb334c86a30c382a2af1b2b332c5/libs/cosmos-sdk/types/context.go#L87

Reply:
That has been fixed, see the latest code.